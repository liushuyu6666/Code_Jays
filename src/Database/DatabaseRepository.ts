import bcrypt from 'bcryptjs';
import mysql, { Connection as MysqlConnection } from 'mysql';
import { MongoClient } from 'mongodb';

// TODO: configurable
const DATABASE_NAME = 'codejays';
const MONGO_URI = 'mongodb://127.0.0.1:27017/CodeJays';

export type DbCallbackFn<T> = (client: MongoClient | MysqlConnection) => Promise<T>;

export type DbOperation = <T>(callback: DbCallbackFn<T>) => Promise<T>;


export enum DatabaseType {
    MongoDB,
    MySQL
} 

export class DatabaseRepository {
    public dbOperation: DbOperation;

    constructor(databaseType: DatabaseType) {
        switch(databaseType) {
            case DatabaseType.MongoDB:
                this.dbOperation = this.mongodbOperation();
                break;
            case DatabaseType.MySQL:
                this.dbOperation = this.mysqlOperation();
                break;
            default:
                throw new Error('Invalid database type');
        }
    }

    private mongodbOperation(): DbOperation {
        return async (callback) => {
            const connection = await MongoClient.connect(MONGO_URI);
            try {
                const result = await callback(connection);
                return result;
            } catch(error) {
                throw error;
            } finally {
                await connection.close();
            }
        };
    }

    private mysqlOperation(): DbOperation {
        return async (callback) => {
            const mysqlConnection: MysqlConnection = mysql.createConnection({
                host: 'localhost',
                user: 'codejays',
                password: 'password',
                database: 'codejays',
            });
            try {
                const result = await callback(mysqlConnection);
                return result;
            } catch (error) {
                throw error;
            } finally {
                mysqlConnection.end();
            }
        }
    }

    /** For Mongodb */
    public async existsCollection(colName: string): Promise<boolean> {
        return await this.dbOperation(async (client) => {
            const col = await (client as MongoClient)
                .db()
                .listCollections({
                    name: colName,
                })
                .toArray();
            return col.length > 0;
        });
    }

    /** For Mongodb */
    public async createCollectionIfNotExists(colName: string): Promise<void> {
        const colExists = await this.existsCollection(colName);
        if (colExists) return;

        await this.dbOperation(async (client) => {
            await (client as MongoClient).db().createCollection(colName);
        });
    }

    public encryptPassword(password: string): string {
        const salt = bcrypt.genSaltSync(10);
        return bcrypt.hashSync(password, salt);
    }

    /** For Mysql */
    public async execSql(sql: string, values: (string | Date)[]): Promise<any> {
        return await this.dbOperation(async (client) => {
            return new Promise((res, rej) => {
                (client as MysqlConnection).query(sql, values, (error, result) => {
                    if (error) rej(error);
                    else res(result);
                });
            });
        });
    }

    /** For Mysql */
    public async existsTable(table: string): Promise<boolean> {
        const sql = `SELECT 1 FROM information_schema.tables WHERE table_schema = ? AND table_name = ? LIMIT 1`;
        const values = [DATABASE_NAME, table];

        const result = await this.execSql(sql, values);
        return result.length > 0;
    }

    /** For Mysql */
    public async createUserTableIfNotExists(): Promise<void> {
        const tableExists = await this.existsTable('user');
        if(tableExists) return;
        const sql = `
            CREATE TABLE user (
                userId VARCHAR(255) PRIMARY KEY,
                username VARCHAR(255) NOT NULL,
                password VARCHAR(255) NOT NULL,
                email VARCHAR(255) NOT NULL
            )
        `;
        await this.execSql(sql, []);
        
        // TODO: logger
        console.log("Create user table in mysql");
    }

    /** For Mysql */
    public async createImageTableIfNotExists(): Promise<void> {
        const tableExists = await this.existsTable('image');
        if (tableExists) return;
        const sql = `
            CREATE TABLE image (
                imageId VARCHAR(255) PRIMARY KEY,
                userId VARCHAR(255) NOT NULL,
                fileName VARCHAR(255) NOT NULL,
                url VARCHAR(255) NOT NULL,
                uploadDate TIMESTAMP NOT NULL
            )
        `;
        await this.execSql(sql, []);

        // TODO: logger
        console.log('Create image table in mysql');
    }
}