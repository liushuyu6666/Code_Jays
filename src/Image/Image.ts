export class Image {
    public imageId: string;
    public fileName: string;
    public url: string;
    public uploadDate: Date;

    constructor(
        imageId: string,
        fileName: string,
        url: string,
        uploadDate: Date,
    ) {
        this.imageId = imageId;
        this.fileName = fileName;
        this.url = url;
        this.uploadDate = uploadDate;
    }
}
