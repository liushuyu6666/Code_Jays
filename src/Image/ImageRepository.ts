import { MongoClient } from 'mongodb';
import { Image } from './Image';
import {
    DatabaseRepository,
    DatabaseType,
} from '../Database/DatabaseRepository';

export interface ImageRepository {
    createImage(
        imageId: string,
        userId: string,
        fileName: string,
        url: string,
        uploadDate: Date,
    ): Promise<Image>;

    listImages(userId: string): Promise<Image[] | undefined>;
}

export class MongodbImageRepository
    extends DatabaseRepository
    implements ImageRepository
{
    constructor() {
        super(DatabaseType.MongoDB);
    }

    async createImage(
        imageId: string,
        userId: string,
        fileName: string,
        key: string,
        uploadDate: Date,
    ): Promise<Image> {
        await this.createCollectionIfNotExists('Image');

        const url = `https://${process.env.AWS_S3_BUCKET}.s3.${process.env.AWS_REGION}.amazonaws.com/${key}`;
        await this.dbOperation(async (client) => {
            const col = (client as MongoClient).db().collection('Image');
            return col.insertOne({
                imageId,
                userId,
                fileName,
                url,
                uploadDate,
            });
        });

        return new Image(imageId, userId, fileName, url, uploadDate);
    }

    // TODO: pagination
    async listImages(userId: string): Promise<Image[] | undefined> {
        if(!await this.existsCollection('Image')) {
            return undefined;
        }

        const imagesWithId = await this.dbOperation(async (client) => {
            return (client as MongoClient)
                .db()
                .collection('Image')
                .find({ userId })
                .toArray();
        });

        if (!imagesWithId) {
            return undefined;
        }

        return imagesWithId.map((image) => {
            return new Image(
                image.imageId,
                userId,
                image.fileName,
                image.url,
                image.uploadDate,
            );
        });
    }
}

export class MysqlImageRepository
    extends DatabaseRepository
    implements ImageRepository
{
    constructor() {
        // TODO: put dbOperation into DatabaseRepository
        super(DatabaseType.MySQL);
    }

    async createImage(
        imageId: string,
        userId: string,
        fileName: string,
        key: string,
        uploadDate: Date,
    ): Promise<Image> {
        await this.createImageTableIfNotExists();

        const url = `https://${process.env.AWS_S3_BUCKET}.s3.${process.env.AWS_REGION}.amazonaws.com/${key}`;
        const sql =
            'INSERT INTO image (imageId, userId, fileName, url, uploadDate) VALUES (?, ?, ?, ?, ?)';
        const values = [imageId, userId, fileName, url, uploadDate];

        await this.execSql(sql, values);

        return new Image(imageId, userId, fileName, url, uploadDate);
    }

    // TODO: pagination
    async listImages(userId: string): Promise<Image[] | undefined> {
        if(!await this.existsTable('image')) {
            return undefined;
        }

        const sql = 'SELECT * FROM image where userId = ?';

        const results = await this.execSql(sql, [userId]);
        if (results) {
            const temp = results.map((result: any) => {
                return new Image(
                    result.imageId,
                    userId,
                    result.fileName,
                    result.url,
                    result.uploadDate,
                );
            });
            return temp;
        } else {
            return undefined;
        }
    }
}

export class ImageRepositoryFactory {
    static createImageRepository(databaseType: DatabaseType): ImageRepository {
        switch (databaseType) {
            case DatabaseType.MongoDB:
                return new MongodbImageRepository();
            case DatabaseType.MySQL:
                return new MysqlImageRepository();
            default:
                throw new Error('Invalid database type');
        }
    }
}
