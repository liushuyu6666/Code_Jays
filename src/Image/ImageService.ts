import { ImageRepository, ImageRepositoryFactory } from './ImageRepository';
import { S3Repository } from '../S3/S3Repository';
import { v4 as uuidv4 } from 'uuid';
import { Image } from './Image';
import { S3Client } from '@aws-sdk/client-s3';
import { DatabaseType } from '../Database/DatabaseRepository';

export class ImageService {
    private imageRepository: ImageRepository;
    private s3Repository: S3Repository;

    constructor(
        databaseType: DatabaseType,
        s3Client: S3Client,
        bucketName: string,
    ) {
        this.imageRepository =
            ImageRepositoryFactory.createImageRepository(databaseType);
        this.s3Repository = new S3Repository(s3Client, bucketName);
    }

    async uploadImage(
        imageName: string,
        imageContent: Express.Multer.File,
    ): Promise<Image | undefined> {
        // Generate id
        const imageId = uuidv4();

        const url = await this.s3Repository.uploadFile(
            imageId,
            imageContent,
            'images',
        );
        if (!url) {
            return undefined;
        }

        return await this.imageRepository.createImage(
            imageId,
            imageName,
            url,
            new Date(),
        );
    }

    async listImages(): Promise<Image[] | undefined> {
        return await this.imageRepository.listImages();
    }
}
