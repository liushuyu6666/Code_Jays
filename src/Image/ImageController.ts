import { DbOperation } from "../../db";
import { DatabaseType } from "../User/UserRepository";
import { ImageService } from "./ImageService";
import { Request, Response } from "express";


export class ImageController {
    private imageService: ImageService;

    constructor(databaseType: DatabaseType, dbOperation: DbOperation, s3Client: AWS.S3, bucketName: string) {
        this.imageService = new ImageService(databaseType, dbOperation, s3Client, bucketName);
    }

    async uploadImageToS3(req: Request, res: Response): Promise<Response<any, Record<string, any>>> {
        const { imageName, imageContent } = req.body;

        if(!imageContent || !imageName) {
            return res
                .status(400)
                .json({ error: "Image or image name are required" });
        }

        await this.imageService.uploadImage(imageName, imageContent);

        return res.status(201).json({ message: `Image ${imageName} uploaded successfully`});
    }
}