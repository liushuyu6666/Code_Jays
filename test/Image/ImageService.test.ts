import { S3Client } from '@aws-sdk/client-s3';
import { DatabaseType } from '../../src/Database/DatabaseRepository';
import { ImageService } from '../../src/Image/ImageService';
import { Image } from '../../src/Image/Image';

const mockImage1 = new Image(
    'imageId',
    'userId',
    'image1',
    'url',
    new Date(Date.parse('2000-01-01')),
);

jest.mock('../../src/Image/ImageRepository', () => {
    return {
        ImageRepositoryFactory: {
            createImageRepository: jest.fn().mockImplementation(() => {
                return {
                    createImage: jest.fn().mockResolvedValue(mockImage1),
                    listImages: jest.fn().mockResolvedValue([mockImage1]),
                };
            }),
        },
    };
});

// TODO: 1, use variable
// TODO: 2, another way to mock S3Repositoryj
jest.mock('../../src/S3/S3Repository', () => {
    return {
        S3Repository: jest.fn().mockImplementation(() => {
            return {
                uploadFile: jest
                    .fn()
                    .mockResolvedValueOnce('url')
                    .mockResolvedValueOnce(undefined),
            };
        }),
    };
});

describe('Test ImageService', () => {
    let imageService: ImageService;
    let mockS3Client: jest.Mocked<S3Client>;

    beforeAll(() => {
        imageService = new ImageService(
            DatabaseType.MySQL,
            mockS3Client,
            'bucket-name',
        );
    });

    afterAll(() => {
        jest.clearAllMocks();
    });

    describe('uploadImage', () => {
        it('should upload an image and create an entry in the repository', async () => {
            const mockFile: Express.Multer.File =
                {} as unknown as Express.Multer.File;

            const mockImage = await imageService.uploadImage(
                'fileName',
                mockFile,
                'userId',
            );

            expect(mockImage).toMatchSnapshot();
        });

        it('should return undefined if url is undefined', async () => {
            const mockFile: Express.Multer.File =
                {} as unknown as Express.Multer.File;

            const mockImage = await imageService.uploadImage(
                'fileName',
                mockFile,
                'userId',
            );

            expect(mockImage).toMatchSnapshot();
        });
    });

    describe('listImage', () => {
        it('should return image array', async () => {
            const mockImage = await imageService.listImages('userId');

            expect(mockImage).toMatchSnapshot();
        });
    });
});
