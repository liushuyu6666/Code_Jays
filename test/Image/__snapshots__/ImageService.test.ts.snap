// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ImageService listImage should return image array 1`] = `
[
  Image {
    "fileName": "image1",
    "imageId": "imageId",
    "uploadDate": 2000-01-01T00:00:00.000Z,
    "url": "url",
    "userId": "userId",
  },
]
`;

exports[`Test ImageService uploadImage should return undefined if url is undefined 1`] = `undefined`;

exports[`Test ImageService uploadImage should upload an image and create an entry in the repository 1`] = `
Image {
  "fileName": "image1",
  "imageId": "imageId",
  "uploadDate": 2000-01-01T00:00:00.000Z,
  "url": "url",
  "userId": "userId",
}
`;
